### PR TITLE
New game room state

### PR DESCRIPTION
**Pull request by shvshv44**

### CODE REVIEW

File: client/src/app/game-room-screen/game-room-screen.component.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  private readonly serverURL = 'http://localhost:8080';
+
+  constructor(public gameRoom: GameRoomService,
+              public webSocket: WebSocketApiService,
+              public messageHandler: MessageHandlerService,
+              private router: Router,
+              private http: HttpClient) {
+    messageHandler.assignHandler('RoomStateResponse', (message) => {
+      this.gameRoom.allPlayers = message.players;
```
Comment (SharonEizner) [inline client/src/app/game-room-screen/game-room-screen.component.ts:23]: The server response with this type of messages

File: client/src/app/game-room-screen/game-room-screen.component.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    this.webSocket._sendRoomStatusRequest(
+      {
+        type: 'RoomStateRequest',
+        gameID: this.gameRoom.gameID
+      });
+  }
+
+  async startGame(): Promise<void> {
+    const startGameURL = this.serverURL + '/StartGame' + '/' + this.gameRoom.gameID;
```
Comment (SharonEizner) [inline client/src/app/game-room-screen/game-room-screen.component.ts:46]: '/StartGame' + '/'  => '/StartGame/'

File: client/src/app/game-room.service.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { Injectable } from '@angular/core';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class GameRoomService {
+
+  public gameID: string;
+  public playerName: string;
+  public allPlayers: string[];
```
Comment (SharonEizner) [inline client/src/app/game-room.service.ts:10]: string[] is id?
if not its need to be id + name

File: client/src/app/web-socket-api.service.ts [hunk 4]
```diff
@@ -1,43 +1,67 @@
...
-  topic = '/topic/gameStatus/';
+  topic = '/topic/gameStatus';
   stompClient: any;
   gameId = '-1';
+  isOpen = false;
 
-  constructor() {
-    this._connect();
+  constructor(public msgHandler: MessageHandlerService) {
+    // this._connect();
```
Comment (SharonEizner) [inline client/src/app/web-socket-api.service.ts:19]: delete this and the function _connect

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/config/WebSocketConfig.java [hunk 5]
```diff
@@ -18,7 +18,7 @@ public void configureMessageBroker(MessageBrokerRegistry config) {
 
     @Override
     public void registerStompEndpoints(StompEndpointRegistry registry) {
-        registry.addEndpoint("/quoridor-websocket").setAllowedOrigins("http://localhost:4200").withSockJS();
+        registry.addEndpoint("/quoridor-websocket").setAllowedOriginPatterns("*").withSockJS();
     }
 
 }
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/config/WebSocketConfig.java:24]: add line

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java [hunk 6]
```diff
@@ -13,48 +22,76 @@
...
+        this.messageSender = messageSender;
     }
 
+    @CrossOrigin
     @GetMapping("/CreateGame/{playerName}")
     @ResponseBody
     public String createGame(@PathVariable String playerName) {
-        return gameRoomManager.createGame(playerName);
+        String gameId = gameRoomManager.createGame(playerName);
+        System.out.println(playerName + " has been created game room with id: " + gameId);
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java:41]: change to logger

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java [hunk 7]
```diff
@@ -13,48 +22,76 @@
...
+            return new ResponseEntity<>(ex.getMessage(), new HttpHeaders(), HttpStatus.BAD_REQUEST);
+        }
     }
 
+    @CrossOrigin
     @GetMapping("/StartGame/{gameId}")
     @ResponseBody
     public String startGame(@PathVariable String gameId) {
         try {
+            System.out.println("Starting game with id: " + gameId);
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java:63]: same change to logger

Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java:63]: did u check that u have more than one player in the game?

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java [hunk 8]
```diff
@@ -13,48 +22,76 @@
...
         } catch (Exception ex) {
             ex.printStackTrace(); // TODO: wont work till TCPPlayer will be implemented!
         }
 
-        return "Game " + gameId + " has been started!";
+        this.messageSender.convertAndSend("/topic/gameStatus/" + gameId, createStartGameEvent(gameId));
+        return gameId;
     }
 
+    @CrossOrigin
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java:73]: what is means? CrossOrigin

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java [hunk 9]
```diff
@@ -13,48 +22,76 @@
...
     @GetMapping("/History/{gameId}")
     @ResponseBody
     public GameResult historyByGameId(@PathVariable String gameId) {
         return historyResolver.getResultByGameId(gameId);
     }
 
+    private StartGameEvent createStartGameEvent(String gameId) {
+        GameRoomState roomState = gameRoomManager.getRoomState(gameId);
+        StartGameEvent startGame = new StartGameEvent();
+        startGame.setType("StartGameEvent");
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameAPIController.java:90]: @Guyeise1 
should be enum
add to the interface gameEvent

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameWebSocket.java [hunk 10]
```diff
@@ -1,23 +1,44 @@
...
+    public GameWebSocket(SimpMessagingTemplate messageSender, GameRoomsManagerService roomsManager, Gson gson) {
+        this.messageSender = messageSender;
+        this.roomsManager = roomsManager;
+        this.gson = gson;
+    }
+
+    @MessageMapping("/hello")
+    @SendTo("/topic/gameStatus")
+    public boolean greeting(String message) throws Exception {
+        System.out.println("sharon");
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameWebSocket.java:38]: logger :)

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameWebSocket.java [hunk 11]
```diff
@@ -30,6 +51,21 @@ public void putWall(@PathVariable String gameId, PlaceWallAction action) {
...
     }
 
+    @MessageMapping("/{gameId}/roomStateRequest")
+    public void roomStateRequest(@DestinationVariable String gameId, String requestAsString) {
+        RoomStateRequest request = gson.fromJson(requestAsString, RoomStateRequest.class);
+        System.out.println("game id " + gameId + " , Input : " + request);
+        roomStateResponse(request);
+    }
+
+    public void roomStateResponse(RoomStateRequest request) {
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/controllers/GameWebSocket.java:61]: didnt @RavivGuy already done this?

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/common/WebSocketMessage.java [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package com.rl.mpquoridor.models.common;
+
+import lombok.Data;
+
+@Data
+public class WebSocketMessage {
+    private String type;
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/common/WebSocketMessage.java:7]: enum

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/common/WebSocketMessage.java [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+    private String gameID;
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/common/WebSocketMessage.java:8]: uuid?

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/gameroom/RoomStateResponse.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package com.rl.mpquoridor.models.gameroom;
+
+import com.rl.mpquoridor.models.common.WebSocketMessage;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+public class RoomStateResponse extends WebSocketMessage {
+    private List<String> players;
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/gameroom/RoomStateResponse.java:10]: what is this?

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/services/GameRoomsManagerService.java [hunk 15]
```diff
@@ -25,13 +26,23 @@ public String createGame(String playerName) {
...
+
+        GameRoomState roomState =  gameRooms.get(gameId);
+        if (roomState.getPlayers().size() < MAX_NUMBER_PLAYERS) {
+            roomState.getPlayers().add(playerName);
+        } else  {
+            throw new InvalidOperationException("Game room is full!");
+        }
     }
 
     public void startGame(String gameId) {
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/services/GameRoomsManagerService.java:48]: add checking of more than one player

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/services/GameRoomsManagerService.java [hunk 16]
```diff
@@ -41,6 +52,10 @@ public void startGame(String gameId) {
         gameManager.run();
     }
 
+    public GameRoomState getRoomState(String gameId) {
+        return gameRooms.get(gameId);
+    }
+
     // TODO: Temporary function till TCP handler will be created!
     private List<Player> createPlayersFromNames(List<String> names) {
         List<Player> players = new ArrayList<>();
```
Comment (SharonEizner) [inline quoridor-logic-server/src/main/java/com/rl/mpquoridor/services/GameRoomsManagerService.java:61]: what?

File: client/src/app/game-room.service.ts (hunk 3 above)
Comment (shvshv44) [reply client/src/app/game-room.service.ts:10]: its list of names

File: quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/common/WebSocketMessage.java (hunk 13 above)
Comment (shvshv44) [reply quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/common/WebSocketMessage.java:8]: no, thids object going to be parsed to json
UUID is object which will be used to make the id

**Review by SharonEizner — APPROVED** [review]

**Review by Guyeise1 — APPROVED** [review]

**Pull request merged**